app/lib/frontend/backend.dart

```diff
@@ -591,53 +597,103 @@ class GCloudPackageRepository extends PackageRepository {
 
   @override
   Future addUploader(String packageName, String uploaderEmail) async {
-    return _withAuthenticatedUser((String userEmail) {
-      return db.withTransaction((Transaction T) async {
-        final packageKey = db.emptyKey.append(models.Package, id: packageName);
-        final models.Package package = (await T.lookup([packageKey])).first;
+    await _withAuthenticatedUser((String userEmail) async {
+      final packageKey = db.emptyKey.append(models.Package, id: packageName);
+      final models.Package package = (await db.lookup([packageKey])).first;
 
-        // Fail if package doesn't exist.
-        if (package == null) {
-          await T.rollback();
-          throw new GenericProcessingException(
-              'Package "$package" does not exist');
-        }
+      _validateNewUploader(userEmail, package, uploaderEmail);
 
-        // Fail if calling user doesn't have permission to change uploaders.
-        if (!package.hasUploader(userEmail)) {
-          await T.rollback();
-          throw new UnauthorizedAccessException(
-              'Calling user does not have permission to change uploaders.');
-        }
+      final status = await backend.updatePackageInvite(
+        packageName: packageName,
+        type: models.PackageInviteType.newUploader,
+        recipientEmail: uploaderEmail,
+        fromEmail: userEmail,
+      );
 
-        // Fail if the uploader we want to add already exists.
-        if (package.hasUploader(uploaderEmail)) {
-          await T.rollback();
-          throw new UploaderAlreadyExistsException();
-        }
+      if (status.isDelayed) {
+        throw new GenericProcessingException(
+            'Previous invite is still active, next notification can be sent '
+            'on ${status.nextNotification.toIso8601String()}.');
+      }
 
-        // Add [uploaderEmail] to uploaders and commit.
-        package.uploaderEmails.add(uploaderEmail);
+      final confirmationUrl = urls.pkgInviteUrl(
+        type: models.PackageInviteType.newUploader,
+        package: packageName,
+        email: uploaderEmail,
+        urlNonce: status.urlNonce,
+      );
+      final message = createUploaderConfirmationEmail(
+        packageName: packageName,
+        activeAccountEmail: userEmail,
+        addedUploaderEmail: uploaderEmail,
+        confirmationUrl: confirmationUrl,
+      );
+      await emailSender.sendMessage(message);
```

Comment:
Don't we need a history event to register that an invitation was sent out?

Reply:
Good point. I guess it could be part of the transaction of the `updatePackageInvite`, or would you like to have it after the email is sent?

Reply:
Before sending the email... ideally, before creation the invitation entry... since dying after we make the invitation just shows that someone tried to send an invitation...

Reply:
Adding it in the transaction where we create/update the invitation entry. With that we will have a separate history entry for each time we wanted to send out the email.